File: abc160/c/rust/c.rs
use std::io::Read;
use std::cmp::max;

fn travel(k: i32, mut a: Vec<i32>) -> i32 {
    let mut m = 0;
    a.push(k + a[0]);
    for w in a.windows(2) {
        m = max(m, w[1] - w[0]);
    }
    k - m
}

fn main(){
    let mut buf = String::new();
    std::io::stdin().read_to_string(&mut buf).unwrap();
    let mut iter = buf.split_whitespace().map(|s| s.parse::<i32>().unwrap());
    let k = iter.next().unwrap();
    let _n = iter.next().unwrap();
    let a: Vec<i32> = iter.collect();

    println!("{}", travel(k, a));
}

File: abc160/c/c.cpp
#include <iostream>
#include <vector>

using namespace std;

int travel(int k, const vector<int> &a) {
  int m = 0;
  int n = a.size();
  for (int i = 0; i < n; i++) {
    if (i != n - 1) {
      m = max(m, a[i + 1] - a[i]);
    } else {
      m = max(m, k - a[i] + a[0]);
    }
  }

  return k - m;
}

int main(void) {
  int k, n;
  cin >> k >> n;

  vector<int> a(n);
  for (int i = 0; i < n; i++) {
    cin >> a[i];
  }
  cout << travel(k, a) << "\n";

  return 0;
}
